### PR TITLE
core/trie: remove trie tracer

### PR DESCRIPTION
**Pull request by holiman**
This PR contains a small portion of the full pbss PR, namely

- Remove the `tracer` from trie (and comitter), and instead using an accessList. 
- Related changes to the `Nodeset`.

### CODE REVIEW

File: trie/nodeset.go [hunk 1]
```diff
@@ -56,104 +54,110 @@ func (n *memoryNode) rlp() []byte {
...
+	}
 	return &NodeSet{
-		owner: owner,
-		updates: &nodesWithOrder{
-			nodes: make(map[string]*nodeWithPrev),
-		},
-		deletes: make(map[string][]byte),
+		owner:      owner,
+		nodes:      make(map[string]*memoryNode),
+		accessList: accessList,
```
Comment (holiman) [inline trie/nodeset.go:108]: This field is actually never used (in this PR)

**Comment by holiman** [comment]
@rjl493456442 did this PR pick an illogical subset of the larger one? Because I can't quite figure out the whole scheme with the accessList. Is it only an optional cache, or something more important?

**Comment by holiman** [comment]
Ah.. so it seems that the trie (Commit) fills entries into the accessList, and then stashes the lookup away inside the nodeset. 
Later on, your difflayers can use the lookup if they need to check the "previous" value for a particular node. 

If that is an accurate description, perhaps a better name would be `priorValues` or `beforeChange` or something ?

File: trie/trie.go [hunk 2]
```diff
@@ -590,7 +568,7 @@ func (t *Trie) Commit(collectLeaf bool) (common.Hash, *NodeSet) {
 		t.root = hashedNode
 		return rootHash, nil
 	}
-	h := newCommitter(t.owner, t.tracer, collectLeaf)
+	h := newCommitter(t.owner, t.accessList, collectLeaf)
 	newRoot, nodes := h.Commit(t.root)
```
Comment (holiman) [inline trie/trie.go:572]: This loop here is a bit wonky. The `t.accessList` is passed to the new `committer`, but the committer ignores it totally, and just passes it along to the `NodeSet`. 
The `h.Commit` operation returns said `NodeSet` in the form of `nodes`. And we deliver `nodes` to the caller. 

It looks a bit odd that we pass the accesslist down into `newComitter` only for that thing to bolt it onto the `nodes`. We could do that ourselves e.g 

```
h = newComitter(t, owner, collecLeaf)
newRoot, nodes := h.Commit(t.root)
nodes.setAccessList(t.accessList)
```
Alternatively, we could handle them separately: 

```
h = newComitter(t, owner, collecLeaf)
newRoot, nodes := h.Commit(t.root)
t.root = newRoot
return rootHash nodes, accessList
```

**Comment by rjl493456442** [comment]
> @rjl493456442 did this PR pick an illogical subset of the larger one? Because I can't quite figure out the whole scheme with the accessList. Is it only an optional cache, or something more important?

AccessList tracks the blob of loaded nodes from trie.Database, these blobs are regarded as *original value before mutation*. AccessList is optional because it *can't* capture all previous value of dirty nodes. 

The special case(it's the ONLY exception) is if the trie was empty and a bunch of nodes are inserted, in this case we won't load any node, but just put the new nodes in the tree. **But there may be dangling nodes with same node path have been left in database previously, the original value of dirty nodes may not be nil**. This case should be handled differently, the logic is here https://github.com/rjl493456442/go-ethereum/blob/pbss/trie/snap_layertree.go#L253

AccessList can avoid most of the disk reads and be beneficial for the performance. If we choose to not keep accessList, then  the entire `original value set of dirty nodes` is constructed in other places by additional disk reads. It's optional to some extend but actually we need to keep it.

File: trie/trie.go (hunk 2 above)
Comment (rjl493456442) [reply trie/trie.go:572]: The second option might not be my preference. `AccessList` or other information about trie might be changed in the future(we may need more/less), put all of them in nodeSet can avoid touching interface too frequently.

I am fine with the first option.

**Review by rjl493456442 — APPROVED** [review]
Lgtm.

**Comment by holiman** [comment]
@rjl493456442 I applied the change I suggested in https://github.com/ethereum/go-ethereum/pull/25963 in both that PR and this one. If it makes your rebase easier, we can just squash the commit. Or if you rebase that one on top of this one, just drop it.

File: trie/trie.go [hunk 3]
```diff
@@ -590,8 +568,9 @@ func (t *Trie) Commit(collectLeaf bool) (common.Hash, *NodeSet) {
 		t.root = hashedNode
 		return rootHash, nil
 	}
-	h := newCommitter(t.owner, t.tracer, collectLeaf)
-	newRoot, nodes := h.Commit(t.root)
+	nodes := NewNodeSet(t.owner).WithAccessList(t.accessList)
```
Comment (rjl493456442) [inline trie/trie.go:571]: You can construct the nodeSet with both owner and accessList, like `NewNodeSet(t.owner, t.accessList)`, so we can get rid of `WithAccessList `

**Review by rjl493456442 — COMMENTED** [review]
Just a nitpick, good to me

File: trie/trie.go (hunk 3 above)
Comment (holiman) [reply trie/trie.go:571]: Ok

Comment (holiman) [reply trie/trie.go:571]: squashed

**Pull request merged**